kingdom_sdk/ports/unit_of_work.py

```diff
@@ -0,0 +1,34 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import Any, Generator
+
+
+class AbstractUnitOfWork(ABC):
+    def __enter__(self) -> AbstractUnitOfWork:
+        return self
+
+    def __exit__(self, *args: Any) -> None:
+        self.rollback()
+
+    def commit(self) -> None:
+        self._commit()
+
+    def rollback(self) -> None:
+        return self._rollback()
+
+    @abstractmethod
+    def _commit(self) -> None:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _rollback(self) -> None:
+        raise NotImplementedError
+
+    @abstractmethod
+    def execute_native_sql(self, statement: str, **params: Any) -> Any:
+        raise NotImplementedError
```

Comment:
the existence of this method its a coupling to SQL dbs, would not be possible to have a abstract alternative for it? Instead of the `UoW` being used by the `DQLInterface` having the opposite, the `UoW` receives a `DQLCommand` (i'm not good with naming things) and executes it?

```python
def query_stuff(arg1, arg2, uow):
    with uow:
        command = ReadStuff(arg1, arg2)
        result = uow.execute_command(command)
    return result
```

Reply:
I forgot about the usage with no-sql dbs, my bad. The way you proposed looks much better, I'll improve the design following it.

Reply:
I decided only rename the `execute_native_sql` to `execute_native_statement`. Putting the uow outside will add code duplication because always will be necessary to create a with context and then to call the uow.execute... so only changing the name to be more generic keeps the code cleaner.